parlai/tasks/anli/agents.py

```diff
@@ -0,0 +1,141 @@
+
+    data_path = os.path.join(
+        opt['datapath'],
+        ANLI,
+        ANLI_PREFIX + ANLI_VERSION,
+        opt['round'],
+        suffix + '.jsonl',
+    )
+
+    return data_path
+
+
+class RoundBaseTeacher(DialogTeacher):
+    @staticmethod
+    def add_cmdline_args(parser):
+        parser = parser.add_argument_group('RoundBase Teacher Args')
+        parser.add_argument(
+            '--dialog-format',
+            type='bool',
+            default=False,
+            help="True if one would like to convert to a dialogue format without special tokens such as 'Premise'"
+            " and 'Hypothesis' (default: False).",
+        )
+
+    def __init__(self, opt, shared=None):
+        opt = copy.deepcopy(opt)
+        opt['round'] = (
+            opt['task'].split(':')[1] if len(opt['task'].split(':')) > 1 else 1
+        )
+        opt['round'] = opt['round'].upper()
+        if not opt['round'] in ANLI_ROUNDS:
+            raise KeyError(f"Undefined task round: {opt['round']}.")
+
+        data_path = _path(opt)
+        opt['datafile'] = data_path
+        self.dialog_format = opt.get('dialog_format', False)
+        self.id = opt['task'].upper()
+        super().__init__(opt, shared)
+
+    def label_candidates(self):
+        if self.dialog_format:
+            return BICLASS_LABELS
+        return MULTINLI_LABELS
+
+    def setup_data(self, path):
+        print('loading: ' + path)
+        with open(path, 'r') as data_file:
+            for pair_line in data_file:
+                pair = json.loads(pair_line)
+                if pair[ANLI_ANSWER_KEY] == '-':
+                    continue
+
+                label_raw = pair[ANLI_ANSWER_KEY]
+                if label_raw in ANLI_LABEL_DICT:
+                    label_raw = ANLI_LABEL_DICT[label_raw]
+
+                question, answer, clas = convert_to_dialogData(
+                    premise_raw=pair[ANLI_PREMISE_KEY],
+                    hypo_raw=pair[ANLI_HYPO_KEY],
+                    answer_raw=label_raw,
+                    dialog_format=self.dialog_format,
+                )
+
+                yield (question, answer, None, clas), True
+
+
+class R1Teacher(RoundBaseTeacher):
+    pass
+
+
+class R2Teacher(RoundBaseTeacher):
+    pass
+
+
+class R3Teacher(RoundBaseTeacher):
+    pass
+
+
+class DefaultTeacher(MultiTaskTeacher):
+    @staticmethod
+    def add_cmdline_args(parser):
+        parser = parser.add_argument_group('ANLI Teacher Args')
+        parser.add_argument(
+            '--dialog-format',
+            type='bool',
+            default=False,
+            help="True if one would like to convert to a dialogue format without special tokens such as 'Premise'"
+            " and 'Hypothesis' (default: False).",
+        )
+
+    def __init__(self, opt, shared=None):
+        anli_tasks = [
+            'anli:r1',
+            'anli:r2',
+            'anli:r3',
+        ]
+        opt = copy.deepcopy(opt)
+        opt['task'] = ','.join(anli_tasks)
```

Comment:
a bit wary of editing opt here, although i see why you did this (seems this is the way it is done in babi, cbt, etc.). my worry is that this would break in the train loop, since during evaluation, we rely on splitting the task string by `,` and evaluate tasks one by one. but i'm not sure... maybe it's fine

thoughts @stephenroller

Reply:
This is the pattern used in all the Multitask teachers already.

It doesn't mess with the train loop because this is a deepcopy of the opt, so the original one stays untouched.

I'd like to migrate away from this, but it's not for this PR.

Reply:
Thanks for pointing this out! Let's mark it for 'next' fix then.